### PR TITLE
Add Firebase auth and Profile model

### PR DESCRIPTION
**Pull request by whirish**
- [x] Adds Firebase authentication via `drf-firebase-auth` to integrate with django auth module
- [x] Adds Profile model to extend django auth's `User` model with location and votes
    - as post-save hook
    - [x] API Endpoint for votes
- [x] Unit tests

### CODE REVIEW

**Comment by whirish** [comment]
All C/U/D API requests must supply `JWT [idToken]` in the Authorization header. Obtaining idToken is described here https://firebase.google.com/docs/auth/admin/verify-id-tokens

File: config/requirements.txt [hunk 1]
```diff
@@ -18,5 +18,6 @@ django-ordered-model~=3.1.1
 django-redis~=4.10.0
 django-debug-toolbar~=1.11
 django-cacheops~=4.1
+git+https://github.com/garyburgmann/drf-firebase-auth.git # Needed until next release
```
Comment (MiklerGM) [inline config/requirements.txt]: Can the bot handle an update or we should update it manually?

Comment (whirish) [reply config/requirements.txt]: I don't think the bot has an option to do that, but I'm watching for new releases on their GitHub so we can do it manually

**Comment by whirish** [comment]
bors t+

File: project/api/models.py [hunk 2]
```diff
@@ -304,3 +307,30 @@ class Narration(OrderedModel):
...
+
+
+class Profile(models.Model):
+    """
+    Optional profile fields, 1-1 with User instances
+    """
+
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    location = models.PointField(blank=True, null=True)
+    votes = JSONField(default=dict, blank=True)
```
Comment (MiklerGM) [inline project/api/models.py]: Do we store votes for narratives and sandbox votes in one field?

Comment (whirish) [reply project/api/models.py]: Good catch, let's split them up and enforce an o2o constraint in the serializer

**Comment by whirish** [comment]
@MiklerGM We could add an int `score` field that's updated as a post_save hook to narration / STVs

**Comment by MiklerGM** [comment]
@whirish can you add more details? If it will work as `score`  for the user, I am approving it :-)

**Comment by whirish** [comment]
Sorry, I was thinking after a vote is cast, we can increment / reduce  a `score` column on the User model corresponding to the target STV or narration's author.

**Comment by MiklerGM** [comment]
let's ask @v1kun 

What for do we need votes, how to store them, how to work with them.

I thought we might want to store use vote in a user profile to prevent double voting on narratives and sandbox.

And for the sandbox, we should keep upvotes and downvotes, or something like this.
I don't have a clear vision here.

There was a prototype of sandbox interface
image

**Comment by v1kun** [comment]
I don't really sure about voting system, some time ago I was thinking about full grown reccomender system, but now i don't think that's needed. I've found [some](https://angularfirebase.com/lessons/reddit-style-upvoting-in-angular-4-and-firebase-nosql/) cool solution for Reddit style voting, but the example is on AngularJS.

**Comment by MiklerGM** [comment]
@v1kun ok, let's drop the idea of giving some recommendations.

We were discussing storing vote data in our db, not in firebase. It's a NoSQL example, so it's not very useful for us.

Could you give us an example what and how should we store for voting on narratives and sandboxed items.

**Comment by v1kun** [comment]
Ok, as far as I know, best practice for storing votes or scores in database is pretty simple, but yet fast. 
Table:
- user_id int
- item_id int
- score bool/int

So for each vote/score, there is one record. I guess you can even store different types of scores in one table (e.g. 0..5 stars or upvote/downvote). 
It's pretty fast on db side for calculating scores by user or by item. 
Examples: We want to calculate overall score for narative, or personal score for users, but it might be tricky to calculate statistics with different scales per user.

I'm not a big fan of using json here, because (de)serialization may cost much more than sql aggregation.

**Comment by MiklerGM** [comment]
@v1kun that means we need more than one additional table for voting
And that was already proposed in https://github.com/chronhq/backend/issues/89

**Comment by whirish** [comment]
I wouldn't be too concerned about performance here, `jsonb` de/serialization in postgres tends to be a pretty fast operation to my knowledge, even with thousands of rows.

**Comment by MiklerGM** [comment]
### What do we have and what was proposed?

#### Create a new table for votes. #89

_Pros:_
1. easy to store
2. easy to fetch
3. everything in one place
4. fast and reliable

_Cons:_
1. django's relation model
2. we should keep votes for narratives and sandbox items in separate tables

#### Store votes in users data #102 

_Pros:_
1. flexible

_Cons:_
1. stale data in JSON
2. additional serialization

### What do we need to cover?

Originally the system was supposed to have all sorts of user votes: for sources, for narratives, for links between sources and actual data, etc.

Now we are trying to solve two separate tasks
1. Voting on narratives
2. Voting in sandbox

For narratives, we need to calculate the total score and protect them from vandalism.
~~In the best case scenario, we should allow voting on narratives without authentication.~~ Probably too much work for us now.
For me, it sounds like the append-only event list.
Probably DB wise we should create a separate table and keep and modular.

For sandbox items...
Well, our goal is to keep votes close to the item, and they are not necessary should be located in the user profile.
- votes would be shown in UI
- based on votes we will calculate the status of sandboxed changes and process data on a schedule.

Let's treat sandboxed items as temporary, not constant data.

Probably we should add JSON field to the sandboxed items 
```json
{ "user_1": true, "user_2": false }
```
Or we could even extend this data and store additional information like data quality or event comments
```json
{ "user_1": { "vote": true, "quality": 5 }, "user_2": { "vote": false, "quality": 2 } }
```

What matters here is the fact that sandboxed item was approved, and if it was successfully merged we should increment rating for the user who opened it.

**Comment by whirish** [comment]
Thoughts on the new table inheritance structure before I update tests? 

New API response, which fields do we need for nested NarrativeVote serialization? Not `narrative` because that's included at the top level, but what others?
```js
    // /api/narratives
    {
        // ...
        "votes": [
            {
                "id": 1,
                "vote": true,
                "user": 1,
                "narrative": 1
            }
        ],
    }
```

File: project/api/models.py [hunk 3]
```diff
@@ -257,6 +309,9 @@ class Narrative(models.Model):
     url = models.TextField(unique=True)
     description = models.TextField()
     tags = ArrayField(models.TextField(max_length=100))
+    votes = models.ManyToManyField(
```
Comment (MiklerGM) [inline project/api/models.py]: Are you sure that this field should have a Many-to-Many relationship?

File: project/api/models.py [hunk 4]
```diff
@@ -19,13 +19,65 @@
...
+
+class NarrativeVote(Vote):
+    """
+    Stores votes for Narratives. Extends Vote model
+    """
+
+    narrative = models.ForeignKey("Narrative", on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ("narrative", "user")
```
Comment (MiklerGM) [inline project/api/models.py]: What will happen if some user decided to change his vote? Will the value be updated?

File: project/api/models.py (hunk 3 above)
Comment (whirish) [reply project/api/models.py]: It's not exactly m2m. Multiple users can vote on multiple narratives, but the same user cannot vote on the same narrative multiple times. A m2m field is the only relation type that allows the through (intermediary) field.

File: project/api/models.py (hunk 4 above)
Comment (whirish) [reply project/api/models.py]: In the ORM we can change  `NarrativeVote.objects.get(user={user}, narrative={narrative}).vote`. I guess we can set up filtering in the API for these parameters

**Comment by whirish** [comment]
@MiklerGM We only have user, narrative, and vote (+/-) atm for it in one table, no MTI is used for the abstract table. Do you  propose a change to this?

File: project/api/models.py (hunk 3 above)
Comment (MiklerGM) [reply project/api/models.py]: Ah, ok, so the user is separated, sounds like django magic, again. Ok then.

File: project/api/models.py (hunk 4 above)
Comment (MiklerGM) [reply project/api/models.py]: I did not get understand, can you rephrase it? What exactly should we filter via API requests?

I was asking about upserting new vote result.

example:
- _
Narrative B - rating 0;
- User-A upvoted Narrative B (submitting `true`)
Narrative B - rating 1;
- User-A downvoted Narrative B (submitting `false`)
Narrative B - rating -1;
- User-A took back his vote (submitting `undefined`)
Narrative B - rating 0;

**Comment by MiklerGM** [comment]
@whirish let's stick to use `+/-` for now at least.

File: project/api/models.py (hunk 4 above)
Comment (whirish) [reply project/api/models.py]: I envisioning this flow for the procedure you outlined:


* User-A upvoted Narrative B (submitting true)
    * `POST` to  `/api/narrative-votes/`
* User-A downvoted Narrative B (submitting false)
    * `PUT` to `/api/narrative-votes/[id of vote in first step]/`
* User-A took back his vote (submitting undefined)
    * `DELETE` to `/api/narrative-votes/[id of vote in first step]/`

And then, to retrieve the ID of the vote, I was thinking we should set up some filtering in the API by `?narrative` and `?user`. How do you feel about this flow?

Comment (MiklerGM) [reply project/api/models.py]: Well, that will require frontend to know about every vote and keep track of them.

I was thinking of less RESTful approach - using `narrative_id` as the key, not the `vote_id`.

* User-A upvoted Narrative B (submitting true)
  * POST to /api/narrative-votes/[narrative_id]/
* User-A downvoted Narrative B (submitting false)
  * POST to /api/narrative-votes/[narrative_id]/
* User-A took back his vote (submitting undefined)
  * POST to /api/narrative-votes/[narrative_id]/

But it is probably because I don't want to think about votes on the frontend

Comment (whirish) [reply project/api/models.py]: Perhaps, as to not break RESTful guidelines too much, we could POST to /api/narrative-votes/ with `{ narrative, user, vote }` and this will update an existing reference if it already exists, using the django helper method [update_or_create](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#update-or-create). I can also configure a delete to be performed if `vote: null`, I think DRF strips `undefined` values in the serializer.

> And it would be great if it will be possible to return new vote results after POSTing vote

To clarify, are you thinking after the request something like `{ upvotes, downvotes }` could be returned for that specific narrative? I'm fine with that, and with returning those counters in /api/narratives/

Comment (MiklerGM) [reply project/api/models.py]: > with { narrative, user, vote }

Ok, that will work

I can fetch `/api/narratives/` after voting

File: auth.env.sample [hunk 5]
```diff
@@ -0,0 +1,2 @@
+TEST_USER_UID=uid
```
Comment (MiklerGM) [inline auth.env.sample]: As far as I understand, this is used only for the unit tests. Probably it would be better to create a new user, and then delete it somewhere in the tests? Otherwise, in case of changing the `firebase.json` or removing this user, the tests will fail.

File: auth.env.sample [hunk 6]
```diff
@@ -0,0 +1,2 @@
...
+CLIENT_API_KEY=key
```
Comment (MiklerGM) [inline auth.env.sample]: This field is not used in production?

File: auth.env.sample (hunk 5 above)
Comment (whirish) [reply auth.env.sample]: Yes it would, I hadn't thought that

File: auth.env.sample (hunk 6 above)
Comment (whirish) [reply auth.env.sample]: No, CLIENT_API_KEY is only needed to retrieve an idToken for the test user. It's not optimal, but I guess it might make sense since the only reason idToken is needed is to verify authentication when communicating with a custom backend. But if anyone knows of a way to retrieve idTokens for tests without needing a client API key, that would be optimal.

File: project/api/serializers.py [hunk 7]
```diff
@@ -34,9 +35,44 @@
...
+    """
+
+    def create(self, validated_data):
+        """
+        Updates NarrativeVote instance if it already exists
+        """
+
+        narrative_vote, _ = NarrativeVote.objects.update_or_create(
+            narrative=validated_data.get("narrative", None),
+            user=validated_data.get("user", None),
```
Comment (MiklerGM) [inline project/api/serializers.py]: That is dangerous, if you are authorized you can send any user_id you want using any id token.

Can you retrieve user data from request headers?
Anyway we don't have any mechanism to fetch user_id (for current user) from the backend, so we can't send this as a payload from frontend.

**Comment by MiklerGM** [comment]
How do we deal with microservices after we merge this PR?

File: project/api/serializers.py [hunk 8]
```diff
@@ -137,6 +173,7 @@ class NarrativeSerializer(ModelSerializer):
 
     start_year = SerializerMethodField()
     end_year = SerializerMethodField()
+    votes = SerializerMethodField()
```
Comment (MiklerGM) [inline project/api/serializers.py]: Can we also add a counter for narrations in narrative?
It might be valuable for showing number of cards in narrative before selection.
One of the old prototypes
image

https://sketch.cloud/s/ZdyEa/a/mqqybk

**Comment by whirish** [comment]
By firebase table are you referring to the user's instance in firebase auth? If so, I believe it already is in sync, other than user deletion. I like the integration with django authentication, even if it does add a bit of complexity, because it allows us to extend the built-in `User` table with extra fields and use their API as well.
How does this PR affect the merging of microservices? Sorry, need to re-orient myself with the project after it slipping past me for a bit.

**Comment by MiklerGM** [comment]
bors try

**Comment by MiklerGM** [comment]
I would propose to close this PR after fixing two things:
- [ ] User token is not checked against header
- [ ] Add a counter for narrations in a narrative (it is unrelated to the PR, but that part was already rewritten)

**Comment by MiklerGM** [comment]
bors try

**Comment by MiklerGM** [comment]
bors try

File: Dockerfile [hunk 9]
```diff
@@ -28,6 +28,7 @@ RUN set -ex \
 RUN mkdir /config
 COPY /config/requirements.txt /config/
 RUN pip install -r /config/requirements.txt
+COPY /config/firebase.json /config/firebase.json
```
Comment (MiklerGM) [inline Dockerfile]: I would propose to mount or pass via ENV this config in the future.
In case we will publish our containers to the docker hub our firebase config would be open to everybody.

File: auth.env.sample (hunk 6 above)
Comment (MiklerGM) [reply auth.env.sample]: I have removed the part with authorization on firebase servers, and this variable is unused now. We can keep it for future integration tests

File: project/api/views.py [hunk 10]
```diff
@@ -112,6 +119,30 @@ class SpacetimeVolumeViewSet(viewsets.ModelViewSet):
...
+    queryset = NarrativeVote.objects.all()
+    serializer_class = NarrativeVoteSerializer
+    permission_classes = (IsAuthenticatedOrReadOnly, IsUserOrReadOnly)
+
+    def create(self, request, *args, **kwargs):
+        """
+        Deletes instance if vote is null
+        """
+
+        request.data["user"] = request.user.id
```
Comment (MiklerGM) [inline project/api/views.py]: Probably we could update the NarrativeVotes model and manage it the same way as Profile
https://github.com/chronhq/backend/blob/bf063e1b37d7dac37e9e2e2531ac5c7913538a67/project/api/models.py#L64-L65

File: project/api/tests/api_tests.py [hunk 11]
```diff
@@ -17,158 +19,101 @@
...
-    Narration,
-    CachedData,
-    City,
-)
+from api.models import PoliticalRelation, CachedData
+
+from .test_data import set_up_data, wiki_cd
+
+
+class FakeUser:  # pylint: disable=too-few-public-methods
```
Comment (MiklerGM) [inline project/api/tests/api_tests.py]: Probably class is not the best way to handle this mock

File: project/api/models.py [hunk 12]
```diff
@@ -139,7 +191,10 @@ class CachedData(models.Model):
 
     history = HistoricalRecords()
 
-    def save(self, *args, **kwargs):  # pylint: disable=W0221
+    def query_wikidata(self):
```
Comment (MiklerGM) [inline project/api/models.py]: This method is mocked by wiki_cd
https://github.com/chronhq/backend/blob/bf063e1b37d7dac37e9e2e2531ac5c7913538a67/project/api/tests/test_data.py#L36-L45
